spring-cloud-gcp-security-firebase/src/main/java/org/springframework/cloud/gcp/security/firebase/FirebaseJwtTokenDecoder.java

```diff
@@ -0,0 +1,190 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.security.firebase;
+
+import java.io.ByteArrayInputStream;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPublicKey;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTParser;
+import com.nimbusds.jwt.SignedJWT;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.oauth2.core.OAuth2TokenValidator;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtException;
+import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
+import org.springframework.web.client.RestOperations;
+
+
+/**
+ * Decodes a Firebase token into a {@link Jwt} token.
+ * This decoder downloads public keys from https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com.
+ * Keys are rotated often, and expiration date is returned as part of a Cache-Control max-age header.
+ * The keys are cached locally and only refreshed when the expiration time is past.
+ * Besides using the RSA keys to validate the token signature, this decoder also uses a pre=configured {@link org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator}
+ * to validate all the claims.
+ * The following validators are used by this class:
+ * {@link org.springframework.security.oauth2.jwt.JwtTimestampValidator} - Validates the expiration date of the Token
+ * {@link org.springframework.security.oauth2.jwt.JwtIssuerValidator} - Validates the iss claim header
+ * {@link FirebaseTokenValidator} - Validates all other headers according to definition at https://firebase.google.com/docs/auth/admin/verify-id-tokens
+ * @author Vinicius Carvalho
+ * @since 1.3
+ */
+public class FirebaseJwtTokenDecoder implements JwtDecoder {
```

Comment:
It won't work, because that is for JWKSet spec, and Firebase does not follow that json standard. It only outputs a dictionary with the `key_id` : ` PEMString`, so the decoder would fail. That's one of the reasons I implemented all from scratch. I have also considered using HttpCachingClient from Apache HttpClient, but it would be able to cache the resource, but the parsing of the certificate would still happen for every token invocation, so I decided to cache the whole operation.

Reply:
Right; sorry -- you've explained this before, and I forgot.